### PR TITLE
Csv generating command

### PR DESCRIPTION
**Pull request by afrasier**
**BEFORE MERGING INTO DEV:**
- [x] Sandbox testing complete 
- [ ] Dev, Stg, Prod S3 buckets and SQS queues are ready
- [ ] IAM roles / other configuration ready for dev, stg, prod

This PR sets up the backend for CSV generation.

* Adds `generate_csv_download` command
  * path - _Required parameter_ - The API path of the data (like `/api/v1/awards` or `/api/v1/references/agency/`)
  * checksum - _Required parameter_ - The checksum of the request to use to filter the queryset.
* Adds `clear_csv` command that deletes all csv files in the s3 bucket. This should be called after a data load, similar to how we currently clear caching
* Adds `clear_request_catalog` command that deletes all RequestCatalog entries. This should be run after as part of the deploy process to ensure no outdated requests hang around in the database
* Adds documentation

**The CSV download endpoint**
Adds `/api/v1/download/` endpoint. This endpoint essentially "converts" any other call into a CSV bulk download call. For example, `/api/v1/download/awards/` or `/api/v1/download/transactions/?req=ABCDdetfg`. You can POST any request that works on the specified endpoint to this endpoint as well. The response from this endpoint looks like:

```
{
  "location": "http://blahblahblah",
  "status": "This file has been requested, and is awaiting queueing",
  "request_checksum": "e78f4722f85",
  "request_path": "/api/v1/awards/"
}
```

* location - The URL where the file can be accessed (once it has been generated)
* status - A plain english description of the current status of the file generation
* request_checksum - The checksum of the request, which can be used to check the status
* request_path - The path of the CSV request

This endpoint checks to  see if the CSV file exists in the configured S3 bucket, and if not, places a message via SQS to have the file generated.

### CODE REVIEW

File: usaspending_api/api_docs/markdown/using_the_api.md [hunk 1]
```diff
@@ -340,6 +340,31 @@ The response has three functional parts:
   * `req` - The special code corresponding to this POST request. See [Post request preservation]("#post-requests-preservation") for how to use this
   * `results` - An array of objects corresponding to the data returned by the specified endpoint. Will _always_ be an array, even if the number of results is only one.
 
+### CSV Bulk Downloads
+
+Bulk CSV downloads are available via the `/api/v1/download/` endpoint. This supports any data endpoint via POST or GET. Because these CSV files can take some time to generate, the response of this endpoint contains the status of file generation and the location of the file once it is complete.
+
+Examples
+* To get a CSV of all Awards, you would access `/api/v1/download/awards/`
+* To get a CSV of all Awards with a particular set of filters, you can pass the request checksum, `/api/v1/download/awards/?req=CHECKSUM` or POST the request object to the endpoint
```
Comment (afrasier) [inline usaspending_api/api_docs/markdown/using_the_api.md]: I will add this

File: usaspending_api/api_docs/markdown/using_the_api.md [hunk 2]
```diff
@@ -340,6 +340,31 @@ The response has three functional parts:
...
+* To check the status of a CSV download, make note of the `request_checksum` and return to the url, with `req=CHECKSUM` attached. For example, `/api/v1/download/awards/?req=CHECKSUM`
+
+Response
+
+```
+{
+  "location": s3://someURL,
+  "status": "This file has been requested, and is awaiting queueing",
```
Comment (afrasier) [inline usaspending_api/api_docs/markdown/using_the_api.md]: Former, descriptions I came up with. We can change them whenever. We can check SQS console to see all requests that are waiting.

File: usaspending_api/common/management/commands/generate_csv_download.py [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+        try:
+            self.logger.info("Beginning generation...")
+            self.logger.info("Destination filename: {}".format(file_name))
+
+            self.logger.info("Resolving path: {}".format(request_path))
+
+            view = resolve_path_to_view(request_path)
+
+            if not view:
+                self.logger.info("Path does not resolve to a currently supported CSV bulk view")
```
Comment (catherinedevlin) [inline usaspending_api/common/management/commands/generate_csv_download.py]: Maybe this should be a higher level than `.info` if it's conking out - I'd call that an `.error`

File: usaspending_api/common/management/commands/generate_csv_download.py [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+import logging
+import sys
+
+from usaspending_api.common.csv_helpers import resolve_path_to_view, create_filename_from_options, format_path
+
+
+class Command(BaseCommand):
+    help = "Generates a CSV file from the specified request checksum and path \
+            and writes the file to the csv_downloads folder with a pre-determined name"
+    logger = logging.getLogger('console')
```
Comment (catherinedevlin) [inline usaspending_api/common/management/commands/generate_csv_download.py]: Nice complete logging on this process.  They're like comments only more useful!

File: usaspending_api/common/management/commands/generate_csv_download.py [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+            # Check the request for a list of requested fields
+            fields = request_catalog.request["data"].get("fields", []) + request_catalog.request["query_params"].get("fields", [])
+
+            if len(fields) > 0:
+                self.logger.info("Requested fields: {}".format(", ".join(fields)))
+                query_set = query_set.values(*fields)
+            else:
+                self.logger.info("No specific fields requested, will render all fields.")
+
+            with open('/'.join([self.CSV_DOWNLOAD_FOLDER_LOCATION, file_name]), 'wb') as csv_file:
```
Comment (catherinedevlin) [inline usaspending_api/common/management/commands/generate_csv_download.py]: I doubt we're actually usable now on Windows, but as a possible improvement to head that way, Python's `os.path` library can join path elements in an OS-independent way (uses backslashes when the OS wants it)

File: usaspending_api/common/models.py [hunk 6]
```diff
@@ -98,6 +100,44 @@ def get_or_create_from_request(request):
...
+            "scope",
+            "usage"
+        ]
+
+        checksumable_request = {
+            "data": {},
+            "query_params": query_params
+        }
+
+        for item in storable_parameters:
```
Comment (catherinedevlin) [inline usaspending_api/common/models.py]: I haven't tried it yet, but I think this could be refactored into

```"data": {k: v for (k, v) in data.items() if k in storable_parameters}```

File: usaspending_api/references/management/commands/clear_request_catalog.py [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+
+    def handle(self, *args, **options):
+        # This will throw an exception and exit the command if the id doesn't exist
+        deleted = RequestCatalog.objects.all().delete()
+
+        statistics = "Statistics:\n  Total objects Removed: " + str(deleted[0])
+        for model in deleted[1].keys():
+            statistics = statistics + "\n  " + model + ": " + str(deleted[1][model])
+
+        self.logger.info(statistics)
```
Comment (catherinedevlin) [inline usaspending_api/references/management/commands/clear_request_catalog.py]: Very nice feedback!

File: usaspending_api/common/csv_helpers.py [hunk 8]
```diff
@@ -0,0 +1,106 @@
...
+    '''
+    sqs = boto3.client('sqs', region_name=settings.SQS_REGION)
+
+    sqs.send_message(QueueUrl=settings.SQS_QUEUE_URL, MessageBody=json.dumps({
+        "request_checksum": checksum,
+        "request_path": format_path(path),
+    }))
+
+
+def format_path(path):
```
Comment (catherinedevlin) [inline usaspending_api/common/csv_helpers.py]: This looks like an ideal function for unit-testing - easy to make tests for, and there are a bunch of corner cases that just might trip it up...

File: usaspending_api/common/csv_helpers.py [hunk 9]
```diff
@@ -0,0 +1,106 @@
...
+        path = "{}/".format(path)
+
+    # Prepend with /api/v1 if it's not there
+    if path[:7] != "/api/v1":
+        path = "/api/v1{}".format(path)
+
+    return path
+
+
+def create_filename_from_options(path, checksum):
```
Comment (catherinedevlin) [inline usaspending_api/common/csv_helpers.py]: As above (and below), a unit test here would be good
